src/base/project.lua

```diff
@@ -314,55 +314,99 @@
 		-- array of paths with should match against that path.
 
-		for _, vpaths in ipairs(prj.vpaths) do
-			for replacement, patterns in pairs(vpaths) do
-				for _, pattern in ipairs(patterns) do
-					local i = abspath:find(path.wildcards(pattern))
-					if i == 1 then
+		local function sort_by_length(a, b)
+			local lena = string.len(a)
+			local lenb = string.len(b)
+			if (lena == lenb) then
+				return a > b
+			end
+			return lena > lenb
+		end
 
-						-- Trim out the part of the name that matched the pattern; what's
-						-- left is the part that gets appended to the replacement to make
-						-- the virtual path. So a pattern like "src/**.h" matching the
-						-- file src/include/hello.h, I want to trim out the src/ part,
-						-- leaving include/hello.h.
+		-- cache patterns.
+		local keys         = prj.cached_vpath_keys
+		local vpaths       = prj.cached_vpaths
+		local wildcards    = prj.cached_wildcards
+
+		if not vpaths then
+			vpaths = {}
+			keys = {}
+			wildcards = {}
+
+			-- flatten vpaths.
+			for _, v in ipairs(prj.vpaths) do
+				for replacement, patterns in pairs(v) do
+					for _, pattern in ipairs(patterns) do
+						vpaths[pattern] = replacement
+					end
+				end
+			end
 
-						-- Find out where the wildcard appears in the match. If there is
-						-- no wildcard, the match includes the entire pattern
+			-- sort keys by length,  we want the longest most unique vpaths first.
+			for pattern, _ in pairs(vpaths) do
+				table.insert(keys, pattern)
+			end
+			table.sort(keys, sort_by_length)
 
-						i = pattern:find("*", 1, true) or (pattern:len() + 1)
+			-- cache wildcard results, path.wildcards is expensive!!!
+			for index, pattern in ipairs(keys) do
+				wildcards[index] = path.wildcards(pattern)
+			end
 
-						-- Trim, taking care to keep the actual file name intact.
+			-- store result.
+			prj.cached_vpath_keys = keys
+			prj.cached_vpaths     = vpaths
+			prj.cached_wildcards  = wildcards
+		end
 
-						local leaf
-						if i < max then
-							leaf = abspath:sub(i)
-						else
-							leaf = fname
-						end
+		-- enumerate vpaths.
+		for index, pattern in ipairs(keys) do
+			local i = abspath:find(wildcards[index])
+			if i == 1 then
 
-						if leaf:startswith("/") then
-							leaf = leaf:sub(2)
-						end
+				-- Trim out the part of the name that matched the pattern; what's
+				-- left is the part that gets appended to the replacement to make
+				-- the virtual path. So a pattern like "src/**.h" matching the
+				-- file src/include/hello.h, I want to trim out the src/ part,
+				-- leaving include/hello.h.
 
-						-- check for (and remove) stars in the replacement pattern.
-						-- If there are none, then trim all path info from the leaf
-						-- and use just the filename in the replacement (stars should
-						-- really only appear at the end; I'm cheating here)
-
-						local stem = ""
-						if replacement:len() > 0 then
-							stem, stars = replacement:gsub("%*", "")
```

Comment:
I think we can optimize this further, because this 'gsub' will always give the exact same answer..